### PR TITLE
[ascend]zcx/support format cast on ascend

### PR DESCRIPTION
**Pull request by zhaochaoxing**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation and Context
Support tensor  convert memory format on ascend.

## Description

- Extents diopiMemoryFormat

- Add format_cast api

## Use cases (Optional)

diopiFormatCast(tensor, diopiMemoryFormat)

## Checklist
**Before PR**:

- [x] I have read and followed the workflow indicated in the [Contributors.md](https://github.com/DeepLink-org/DIOPI/blob/main/Contributors.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [Contributors.md](https://github.com/DeepLink-org/DIOPI/blob/main/Contributors.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: proto/include/diopi/diopirt.h [hunk 1]
```diff
@@ -139,6 +165,8 @@ extern DIOPI_RT_API diopiError_t diopiGetStream(diopiContextHandle_t ctx, diopiS
 
 extern DIOPI_RT_API diopiError_t diopiRequireTensor(diopiContextHandle_t ctx, diopiTensorHandle_t* tensor, const diopiSize_t* size, const diopiSize_t* stride,
                                                     const diopiDtype_t dtype, const diopiDevice_t device);
+extern DIOPI_RT_API diopiError_t diopiRequireTensorWithNumel(diopiContextHandle_t ctx, diopiTensorHandle_t* tensor, const diopiSize_t* size,
```
Comment (zhaoguochun1995) [inline proto/include/diopi/diopirt.h:168]: 这个接口似乎不是很有必要？ 这部分是否精简一点好吧？

File: proto/include/diopi/diopirt.h [hunk 2]
```diff
@@ -85,12 +85,34 @@ typedef struct {
...
+    Undefined = -1,
+    Contiguous = 0,
+    ChannelsLast = 1,
+    ChannelsLast3d = 2,
+    Preserve = 3,
+    ChannelsLast1d = 4,
+    NCHW = 5,
+    NC1HWC0 = 6,
+    FRACTAL_Z = 7,
+    NC1HWC0_C04 = 8,
```
Comment (zhaoguochun1995) [inline proto/include/diopi/diopirt.h:98]: 这里有些是华为特有的memformat，放在这个头文件似乎不太合适？

File: proto/include/diopi/diopirt.h (hunk 1 above)
Comment (zhaochaoxing) [reply proto/include/diopi/diopirt.h]: 已精简

File: impl/ascend/common/acloprunner.hpp [hunk 3]
```diff
@@ -393,6 +394,44 @@ class AclOpRunner {
         return *this;
     }
 
+    AclOpRunner& addInput(const AscendTensor& at, const diopiMemoryFormat_t diopiFormat) {
```
Comment (zhaoguochun1995) [inline impl/ascend/common/acloprunner.hpp:397]: 这个函数和上面这个函数重复度较高，可以优化一下吧

File: impl/ascend/common/acloprunner.hpp [hunk 4]
```diff
@@ -393,6 +394,44 @@ class AclOpRunner {
...
+
+        std::vector<int64_t> dims = at.getAclMemShape();
+        std::vector<int64_t> storageDims = at.getStorageDims();
+        aclFormat baseFormat = FormatHelper::getAclFormat(FormatHelper::getBaseFormat(diopiFormat));
+        desc = aclCreateTensorDesc(dtypeCastStrategy(at.dtype()), dims.size(), dims.data(), baseFormat);
+        ASCEND_CHECK_ABORT(desc != nullptr, "aclTensorDesc should not be nullptr.");
+        aclSetTensorFormat(desc, FormatHelper::getAclFormat(diopiFormat));
+        aclSetTensorShape(desc, storageDims.size(), storageDims.data());
+        int64_t bufferSize = at.getStorageNumel() * at.elemsize();
+        buffer = aclCreateDataBuffer(const_cast<void*>(at.data()), bufferSize);
```
Comment (zhaoguochun1995) [inline impl/ascend/common/acloprunner.hpp:418]: 这个地方会有内存泄漏

Comment (zhaochaoxing) [reply impl/ascend/common/acloprunner.hpp]: 已沟通，不存在内存泄漏。

File: proto/include/diopi/diopirt.h [hunk 5]
```diff
@@ -128,6 +151,10 @@ extern DIOPI_RT_API DIOPI_ATTR_WEEK diopiError_t diopiGetTensorDevice(diopiConst
 extern DIOPI_RT_API diopiError_t diopiGetTensorNumel(diopiConstTensorHandle_t th, int64_t* numel);
 extern DIOPI_RT_API DIOPI_ATTR_WEEK diopiError_t diopiGetTensorElemSize(diopiConstTensorHandle_t th, int64_t* itemsize);
 extern DIOPI_RT_API DIOPI_ATTR_WEEK diopiError_t diopiGetTensorStoragePtr(diopiConstTensorHandle_t th, void** pStoragePtr);
+extern DIOPI_RT_API diopiError_t diopiTensorHasStorageDesc(diopiConstTensorHandle_t pth, bool* result);
```
Comment (zhaoguochun1995) [inline proto/include/diopi/diopirt.h:154]: diopiStorageDesc_t 这个在一个tensor 里是可有可无的，还是一定要存在呢？或者说，只在华为设备上需要？memformat里涉及填充的都需要这个吧？

**Review by zhaoguochun1995 — APPROVED** [review]

File: proto/include/diopi/diopirt.h (hunk 5 above)
Comment (zhaochaoxing) [reply proto/include/diopi/diopirt.h]: 去掉了

**Review by hellozmz — APPROVED** [review]

File: impl/ascend/ascend_tensor.hpp [hunk 6]
```diff
@@ -101,6 +104,15 @@ class AscendTensor final {
...
             diopiGetTensorNumel(tensor_, &numel_);
             diopiGetTensorElemSize(tensor_, &elemsize_);
+
+            if (device == diopiDevice_t::diopi_device) {
+                diopiStorageDesc_t desc;
+                diopiGetTensorStorageDesc(tensor_, &desc);
+                storageFormat_ = desc.format;
+                std::vector<int64_t> tmp(desc.sizes.data, desc.sizes.data + desc.sizes.len);
+                storageDims_ = std::move(tmp);
+                storageNumel_ = std::accumulate(storageDims_.begin(), storageDims_.end(), static_cast<int64_t>(1), std::multiplies<>());
```
[the remark this answers is not in the document]
Comment (zhaochaoxing) [reply impl/ascend/ascend_tensor.hpp]: 好的

File: impl/ascend/ascend_tensor.hpp [hunk 7]
```diff
@@ -164,6 +176,24 @@ class AscendTensor final {
         return numel_;
     }
 
+    std::vector<int64_t> storageDims() const {
+        ASCEND_CHECK_NULLPTR_ABORT(tensor_);
+        if (device_ == diopiDevice_t::diopi_device && isContiguous() && dim() > 0) {
```
[the remark this answers is not in the document]
Comment (zhaochaoxing) [reply impl/ascend/ascend_tensor.hpp]: 好的，已删除，只保留device_ == diopiDevice_t::diopi_device

File: impl/ascend/common/format_helper.cpp [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+// NCDHW to NCDHW
+Shape NCDHWFormatInfo::inferShape(const Shape& dims) {
+    ASCEND_CHECK_ABORT(dims.size() == 5, "cannot convert to NCDHW");
+    return dims;
+}
+
+// NCDHW to NDC1HWC0
+Shape NDC1HWC0FormatInfo::inferShape(const Shape& dims) {
+    ASCEND_CHECK_ABORT(dims.size() == 5, "cannot convert to NDC1HWC0");
+    Shape res;
```
[the remark this answers is not in the document]
Comment (zhaochaoxing) [reply impl/ascend/common/format_helper.cpp]: done

File: diopi_test/diopi_stub/include/litert.hpp [hunk 9]
```diff
@@ -63,6 +66,28 @@ class Storage final {
         nbytes_ = 0;
     }
 
+    void initDesc(const std::vector<int64_t>& shape) {
+        storageSizes_ = shape;
+        if (shape.size() == 5) {
+            format_ = diopiMemoryFormat_t::NCDHW;
```
[the remark this answers is not in the document]
Comment (zhaochaoxing) [reply diopi_test/diopi_stub/include/litert.hpp]: 已改正

File: diopi_test/diopi_stub/csrc/litert.cpp [hunk 10]
```diff
@@ -187,6 +189,11 @@ diopiTensor& diopiTensor::operator=(const diopiTensor& other) {
     return *this;
 }
 
+void diopiTensor::copyMetaData(const diopiTensor& other) {
```
Comment (yangbofun) [inline diopi_test/diopi_stub/csrc/litert.cpp:192]: 这里的名叫copyMetaData, 实现里为何只复用了shape_和stride。

File: impl/ascend/ascend_tensor.cpp [hunk 11]
```diff
@@ -147,7 +148,7 @@ std::vector<int64_t> AscendTensor::getAclMemShape() const {
 int64_t AscendTensor::getAclMemBufferSize() const {
     if (this->isContiguous()) {
         if (dim() > 0) {
-            return this->numel() * this->elemsize();
+            return this->storageNumel() * this->elemsize();
```
Comment (yangbofun) [inline impl/ascend/ascend_tensor.cpp:151]: 这个地方可以收敛下，只调用this->storageNumel么？对于dim==0,直接写成1*this->elemsize();

File: impl/ascend/common/format_helper.cpp [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+        dim.emplace_back(1);
+    }
+    if (dim.size() == 1) {
+        dim.emplace_back(1);
+    }
+    int i = 0;
+    for (; i < dim.size() - 2; i++) {
+        res.emplace_back(dim[i]);
+    }
+    res.emplace_back((dim[i + 1] + 15) / blocksize);
```
Comment (yangbofun) [inline impl/ascend/common/format_helper.cpp:175]: 15给为blocksize-1.

File: impl/ascend/common/format_helper.h [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+    Shape inferShape(const Shape& dims) override;
+};
+class ZFormatInfo : public FormatInfo {
+public:
+    ZFormatInfo() : FormatInfo(diopiMemoryFormat_t::FRACTAL_Z, ACL_FORMAT_FRACTAL_Z, ACL_FORMAT_NCHW, "FRACTAL_Z", true) {}
+    Shape inferShape(const Shape& dims) override;
+};
+class NDHWCFormatInfo : public FormatInfo {
+public:
+    NDHWCFormatInfo() : FormatInfo(diopiMemoryFormat_t::NDHWC, ACL_FORMAT_NDHWC, ACL_FORMAT_NCDHW, "NDHWC", false) {}
```
[the remark this answers is not in the document]
Comment (zhaochaoxing) [reply impl/ascend/common/format_helper.h:96]: 没错，ACL_FORMAT_NDHWC的baseFormat是ACL_FORMAT_NCDHW

File: impl/ascend/common/format_helper.cpp (hunk 12 above)
Comment (zhaochaoxing) [reply impl/ascend/common/format_helper.cpp:175]: 好的

File: diopi_test/diopi_stub/csrc/litert.cpp (hunk 10 above)
Comment (zhaochaoxing) [reply diopi_test/diopi_stub/csrc/litert.cpp:192]: 新增了numel，我看liter tensor里的storage_offset固定是0，不用拷贝。

**Review by yangbofun — APPROVED** [review]

**Comment by yangbofun** [comment]
我先revert掉，等dipu和diopi都就绪后，再合并。

**Pull request merged**